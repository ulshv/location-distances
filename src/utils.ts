import uniq from 'lodash/uniq';

export const getLowercasedSortedList = (items: string[]) => {
  return items.map(l => l.toLowerCase().trim()).sort()
};

export const getListWithUniqueItems = (items: string[]) => {
  return uniq(items);
}

export const parseRequestData = (locations: string[]) => {
  return getListWithUniqueItems(getLowercasedSortedList(locations));
}

export const createPairsFromList = <T>(items: T[]) => {
  const result = [];

  for (let i = 0; i < items.length - 1; i++) {
    for (let j = 1 + i; j < items.length; j++) {
      result.push([items[i], items[j]]);
    }
  }

  return result;
}
