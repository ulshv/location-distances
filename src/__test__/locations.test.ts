import mapApiMock from "./map-api-mock";
import { createPairsFromList } from "../utils";
import {
  _parseRawLocationQueries,
  _fetchLocations,
  _getLocationPairsWithDistances,
  _getClosestLocationPairsWithDistances,
  _formatFinalResult,
  calculateResponse
} from "../locations";


describe('locations.ts', () => {
  it('parseRawLocationQueries() should propertly parse user request data (list of locations)', () => {
    const data = ['San Francisco', 'New York', 'san francisco', 'new york    '];
    const result = ['new york', 'san francisco'];
    expect(_parseRawLocationQueries(data)).toEqual(result);
  })
})


describe('locations.ts', () => {
  const locationsQueryRaw = ['New York', '  MoScoW ', 'PhiladelphiA', 'Saint Petersburg', 'Nowhere'];

  let locationQueries;
  let locations;
  let locationPairs;
  let locationPairsWithDistances;
  let closestLocationPairsWithDistances;
  let formattedResult;

  it('_parseRawLocationQueries() should propertly parse raw data', () => {
    locationQueries = _parseRawLocationQueries(locationsQueryRaw);
    expect(locationQueries).toMatchSnapshot();
  });

  it('_fetchLocations() should fetch data with mock api', async () => {
    locations = await _fetchLocations(mapApiMock, locationQueries)
    expect(locations).toMatchSnapshot();
  });

  it('_getLocationPairsWithDistances() add distance for each location pair', async () => {
    locationPairs = createPairsFromList(locations);
    locationPairsWithDistances = _getLocationPairsWithDistances(locationPairs)
    expect(locationPairsWithDistances).toMatchSnapshot();
  });

  it('_getClosestLocationPairsWithDistances() should find closest pair for each location', async () => {
    closestLocationPairsWithDistances = _getClosestLocationPairsWithDistances(
      locations,
      locationPairsWithDistances
    )
    expect(closestLocationPairsWithDistances).toMatchSnapshot();
  });

  it('_formatFinalResult() should propertly format results', async () => {
    formattedResult = _formatFinalResult(
      locations,
      closestLocationPairsWithDistances
    )
    expect(formattedResult).toMatchSnapshot();
  });

  it('calculateResponse() should propertly transform raw locations queries to final result', async () => {
    const result = await calculateResponse(mapApiMock, locationsQueryRaw);
    expect(result).toMatchSnapshot();
  })

});
