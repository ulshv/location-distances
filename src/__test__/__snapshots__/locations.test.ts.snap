// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`locations.ts _fetchLocations() should fetch data with mock api 1`] = `
Array [
  Object {
    "latitude": 55.75583,
    "longitude": 37.61778,
    "name": "Moscow",
    "query": "moscow",
  },
  Object {
    "latitude": 40.7648,
    "longitude": -73.9808,
    "name": "New York",
    "query": "new york",
  },
  Object {
    "latitude": 40.0115,
    "longitude": -75.1327,
    "name": "Philadelphia",
    "query": "philadelphia",
  },
  Object {
    "latitude": 59.95,
    "longitude": 30.31667,
    "name": "Saint Petersburg",
    "query": "saint petersburg",
  },
]
`;

exports[`locations.ts _formatFinalResult() should propertly format results 1`] = `
Array [
  Object {
    "distance": "635.70 km",
    "from": Object {
      "latitude": 55.75583,
      "longitude": 37.61778,
      "name": "Moscow",
      "query": "moscow",
    },
    "to": Object {
      "latitude": 59.95,
      "longitude": 30.31667,
      "name": "Saint Petersburg",
      "query": "saint petersburg",
    },
  },
  Object {
    "distance": "128.73 km",
    "from": Object {
      "latitude": 40.7648,
      "longitude": -73.9808,
      "name": "New York",
      "query": "new york",
    },
    "to": Object {
      "latitude": 40.0115,
      "longitude": -75.1327,
      "name": "Philadelphia",
      "query": "philadelphia",
    },
  },
  Object {
    "distance": "128.73 km",
    "from": Object {
      "latitude": 40.0115,
      "longitude": -75.1327,
      "name": "Philadelphia",
      "query": "philadelphia",
    },
    "to": Object {
      "latitude": 40.7648,
      "longitude": -73.9808,
      "name": "New York",
      "query": "new york",
    },
  },
  Object {
    "distance": "635.70 km",
    "from": Object {
      "latitude": 59.95,
      "longitude": 30.31667,
      "name": "Saint Petersburg",
      "query": "saint petersburg",
    },
    "to": Object {
      "latitude": 55.75583,
      "longitude": 37.61778,
      "name": "Moscow",
      "query": "moscow",
    },
  },
]
`;

exports[`locations.ts _getClosestLocationPairsWithDistances() should find closest pair for each location 1`] = `
Array [
  Object {
    "distance": 635700,
    "pair": Array [
      Object {
        "latitude": 55.75583,
        "longitude": 37.61778,
        "name": "Moscow",
        "query": "moscow",
      },
      Object {
        "latitude": 59.95,
        "longitude": 30.31667,
        "name": "Saint Petersburg",
        "query": "saint petersburg",
      },
    ],
  },
  Object {
    "distance": 128727,
    "pair": Array [
      Object {
        "latitude": 40.7648,
        "longitude": -73.9808,
        "name": "New York",
        "query": "new york",
      },
      Object {
        "latitude": 40.0115,
        "longitude": -75.1327,
        "name": "Philadelphia",
        "query": "philadelphia",
      },
    ],
  },
  Object {
    "distance": 128727,
    "pair": Array [
      Object {
        "latitude": 40.7648,
        "longitude": -73.9808,
        "name": "New York",
        "query": "new york",
      },
      Object {
        "latitude": 40.0115,
        "longitude": -75.1327,
        "name": "Philadelphia",
        "query": "philadelphia",
      },
    ],
  },
  Object {
    "distance": 635700,
    "pair": Array [
      Object {
        "latitude": 55.75583,
        "longitude": 37.61778,
        "name": "Moscow",
        "query": "moscow",
      },
      Object {
        "latitude": 59.95,
        "longitude": 30.31667,
        "name": "Saint Petersburg",
        "query": "saint petersburg",
      },
    ],
  },
]
`;

exports[`locations.ts _getLocationPairsWithDistances() add distance for each location pair 1`] = `
Array [
  Object {
    "distance": 7512724,
    "pair": Array [
      Object {
        "latitude": 55.75583,
        "longitude": 37.61778,
        "name": "Moscow",
        "query": "moscow",
      },
      Object {
        "latitude": 40.7648,
        "longitude": -73.9808,
        "name": "New York",
        "query": "new york",
      },
    ],
  },
  Object {
    "distance": 7636967,
    "pair": Array [
      Object {
        "latitude": 55.75583,
        "longitude": 37.61778,
        "name": "Moscow",
        "query": "moscow",
      },
      Object {
        "latitude": 40.0115,
        "longitude": -75.1327,
        "name": "Philadelphia",
        "query": "philadelphia",
      },
    ],
  },
  Object {
    "distance": 635700,
    "pair": Array [
      Object {
        "latitude": 55.75583,
        "longitude": 37.61778,
        "name": "Moscow",
        "query": "moscow",
      },
      Object {
        "latitude": 59.95,
        "longitude": 30.31667,
        "name": "Saint Petersburg",
        "query": "saint petersburg",
      },
    ],
  },
  Object {
    "distance": 128727,
    "pair": Array [
      Object {
        "latitude": 40.7648,
        "longitude": -73.9808,
        "name": "New York",
        "query": "new york",
      },
      Object {
        "latitude": 40.0115,
        "longitude": -75.1327,
        "name": "Philadelphia",
        "query": "philadelphia",
      },
    ],
  },
  Object {
    "distance": 6886947,
    "pair": Array [
      Object {
        "latitude": 40.7648,
        "longitude": -73.9808,
        "name": "New York",
        "query": "new york",
      },
      Object {
        "latitude": 59.95,
        "longitude": 30.31667,
        "name": "Saint Petersburg",
        "query": "saint petersburg",
      },
    ],
  },
  Object {
    "distance": 7010527,
    "pair": Array [
      Object {
        "latitude": 40.0115,
        "longitude": -75.1327,
        "name": "Philadelphia",
        "query": "philadelphia",
      },
      Object {
        "latitude": 59.95,
        "longitude": 30.31667,
        "name": "Saint Petersburg",
        "query": "saint petersburg",
      },
    ],
  },
]
`;

exports[`locations.ts _parseRawLocationQueries() should propertly parse raw data 1`] = `
Array [
  "moscow",
  "new york",
  "nowhere",
  "philadelphia",
  "saint petersburg",
]
`;

exports[`locations.ts calculateResponse() should propertly transform raw locations queries to final result 1`] = `
Array [
  Object {
    "distance": "635.70 km",
    "from": Object {
      "latitude": 55.75583,
      "longitude": 37.61778,
      "name": "Moscow",
      "query": "moscow",
    },
    "to": Object {
      "latitude": 59.95,
      "longitude": 30.31667,
      "name": "Saint Petersburg",
      "query": "saint petersburg",
    },
  },
  Object {
    "distance": "128.73 km",
    "from": Object {
      "latitude": 40.7648,
      "longitude": -73.9808,
      "name": "New York",
      "query": "new york",
    },
    "to": Object {
      "latitude": 40.0115,
      "longitude": -75.1327,
      "name": "Philadelphia",
      "query": "philadelphia",
    },
  },
  Object {
    "distance": "128.73 km",
    "from": Object {
      "latitude": 40.0115,
      "longitude": -75.1327,
      "name": "Philadelphia",
      "query": "philadelphia",
    },
    "to": Object {
      "latitude": 40.7648,
      "longitude": -73.9808,
      "name": "New York",
      "query": "new york",
    },
  },
  Object {
    "distance": "635.70 km",
    "from": Object {
      "latitude": 59.95,
      "longitude": 30.31667,
      "name": "Saint Petersburg",
      "query": "saint petersburg",
    },
    "to": Object {
      "latitude": 55.75583,
      "longitude": 37.61778,
      "name": "Moscow",
      "query": "moscow",
    },
  },
]
`;
